sql-plugin/src/main/scala/com/nvidia/spark/rapids/AbstractGpuJoinIterator.scala

```diff
@@ -133,8 +134,13 @@ abstract class AbstractGpuJoinIterator(
   private def nextCbFromGatherer(): Option[ColumnarBatch] = {
     withResource(new NvtxWithMetrics(gatherNvtxName, NvtxColor.DARK_GREEN, joinTime)) { _ =>
       val ret = gathererStore.map { gather =>
-        val nextRows = JoinGatherer.getRowsInNextBatch(gather, targetSize)
-        gather.gatherNext(nextRows)
+        gather.checkpoint
+        withRetryNoSplit[ColumnarBatch] {
+          gather.allowSpilling()
```

Comment:
this has to be outside of `withRetryNoSplit`, because if we fail to contiguous split within the block, we are likely in a bad state (the LazySpillable cached batch is likely going to get closed, since `SpillableColumnarBatch.apply` says it takes ownership).

Reply:
Thanks.   This is where LSCB is problematic.  I think I may need it both before the block and within the block.  I was hoping I could get away with just moving it inside.  The reason I think I need it inside the block as well is to handle closing the `cached` cb that was created when a previous attempt called `getBatch`.  Perhaps the inside the block allowSpilling call should be in a try..catch block.

Reply:
I see that `gatherNext` (within the retry block) calls `getBatch` which leaves around that ColumnarBatch in the lazy spillable. And what we want is the withRetry code to be able to rewind and remove all GPU memory used.

So within `gatherNext` one, hacky, approach would be to:

```
var batch: ColumnarBatch = null
try { 
batch = data.getBatch
val gatheredTable = withResource(GpuColumnVector.form(batch)) { table =>
  table.gather(..)
}
  ..
} catch {
  case t: Throwable => 
    data.dropCache() // new method that just closes the cached copy, and sets the optional to None if something was cached (e.g. it wasn't spillable).
}
```

But this is something we'd want to clean up once everything is spillable by default.